src/odemis/driver/focustracker.py

```diff
@@ -21,206 +21,368 @@
+            node = FakeRemoteNode(node_idx, datasheet)
         else:
-            self.node = canopen.RemoteNode(node_idx, object_dict)
-        self.network.add_node(self.node)
-        self._swVersion = "python-canopen v%s , python-can v%s" % (canopen.__version__, can.__version__)
-        rev_num = self.node.sdo["Identity Object"]["Revision number"].raw
-        minor = rev_num & 0xffff
-        major = (rev_num >> 16) & 0xffff
-        self._hwVersion = "{}.{}".format(major, minor)
-        # Create SDO communication objects to communicate
-        self._position_sdo = self.node.sdo["AI Input PV"][1]
-        self._target_pos_sdo = self.node.sdo["CO Set Point W"][1]
-
-        # Read PID gains from the device (and set the current metadata)
-        self._proportional_gain_sdo = self.node.sdo['CO Proportional Band Xp1'][1]
-        self._integral_gain_sdo = self.node.sdo['CO Integral Action Time Tn1'][1]
-        self._derivative_gain_sdo = self.node.sdo['CO Derivative Action Time Tv1'][1]
-
-        self._tracking_sdo = self.node.sdo['Controller On/ Off'][1]
-        # set VAs only if there is hardware, channel = '' if no hardware
-        self._metadata[model.MD_GAIN_P] = self._proportional_gain_sdo.raw
-        self._metadata[model.MD_GAIN_I] = self._integral_gain_sdo.raw
-        self._metadata[model.MD_GAIN_I] = self._derivative_gain_sdo.raw
-        self.targetPosition = model.FloatContinuous(self._get_target_pos(), TARGET_POSITION_RANGE, unit="m",
-                                                    getter=self._get_target_pos, setter=self._set_target_pos)
-        self.position = model.FloatVA(self._get_position(), unit="m", readonly=True, getter=self._get_position)
-        self.tracking = model.BooleanVA(self._get_tracking(), getter=self._get_tracking,
-                                        setter=self._set_tracking)
+            node = canopen.RemoteNode(node_idx, datasheet)
+        # Note: add_node() supports a "upload_eds" flag to read the object dict from
+        # the device. However the current firmware doesn't support that.
+        network.add_node(node)
+
+        # Check the device is there, and also force the state to be updated
+        try:
+            if channel != "fake":
+                node.nmt.wait_for_heartbeat(timeout=5)
+        except NmtError:
+            raise model.HwError("Focus tracker not found on channel %s with ID %s" % (channel, node_idx))
+
+        logging.debug("Device is in state %s", node.nmt.state)
+
+        # If the device is stopped, it won't answer any SDO
+        if node.nmt.state not in ("OPERATIONAL", "PRE-OPERATIONAL"):
+            node.nmt.state = "PRE-OPERATIONAL"
+            logging.debug("Turning on the device to state %s", node.nmt.state)
+
+        # Check that the device has the right Vendor ID and Product code, mostly
+        # in case the node index corresponds to a different device, also on the network.
+        vid = node.sdo["Identity object"]["Vendor-ID"].raw
```

Comment:
The .eds file now does not contain a "Vendor-ID", shouldn't it have that for this to work?